### PR TITLE
Drill-down example how to find all built-in reserved keywords

### PR DESCRIPTION
**Pull request by dotysan**
Maybe a bit too advanced this early in the course. But still, an optional, fun drill-down experiment using Python to analyze Python.

### CODE REVIEW

File: book/python/02_variables.ipynb [hunk 1]
```diff
@@ -152,9 +154,9 @@
     "        keywords.append(builtin_name)\n",
     "\n",
     "print(f\"Constants: {' '.join(constants)}\")\n",
-    "print(\"Exceptions:\", ' '.join(exceptions['other']))\n",
-    "print(\"Warnings:\", ' '.join(exceptions['warnings']))\n",
-    "print(\"Errors:\", ' '.join(exceptions['errors']))\n",
+    "print(\"Exceptions:\", \" \".join(exceptions[\"other\"]))\n",
```
Comment (dotysan) [inline book/python/02_variables.ipynb:157]: Yuk! Single quoted strings are fine. Especially inside of double-quites.

Comment (dotysan) [reply book/python/02_variables.ipynb]: This appears to be caused by Black's pedantism.

Comment (dotysan) [reply book/python/02_variables.ipynb]: Fixed with 0d333b42.

**Comment by dotysan** [comment]
I do agree that some string/quote cleanup is needed.

However Black's is too pedantic. https://github.com/psf/black/issues/118 Nor do they give you fine grain control over which types are good (docstring, f-string, etc.) and which are bad (simple constants, strings in strings, etc.) IMHO.

So I disabled that feature in Black with --skip-string-normalization.

For the other good string/quote cleanup with finer-grained control, we could use flake8, ruff, or maybe pylint.

**Comment by giswqs** [comment]
Thank you for the pull request! I have some reservations about merging it. The added code block may be a bit too advanced for the first Python lecture. Since we've already covered some essential keywords, I believe those should be sufficient for beginners.

Regarding the Black formatter, I understand your frustration with the lack of a single quotes option. I felt the same way when I first started using it a few years ago. However, I've since adapted to the double quotes convention, and I believe standardization is beneficial. I would prefer to implement this as a pre-commit hook.

**Comment by dotysan** [comment]
I totally agree with your first comment. I'm new to learning notebooks and was hoping it would remain collapsed--and less obtrusive. But alas, I've now learned that it expands the entire code block when you re-open the notebook. And fixing this requires additional plugins and other jazz which just complicates things. Maybe I'll create an asterisk/link to the more in-depth example in a separate notebook...

As for the Black pendanticism, I disagree and prefer leaving room for some sense of style, rather than strict septic standardization. Code is language. And can be expressive too. I.e. `exceptions['errors']` since the key is effectively static, but `print(f"Constants: {' '.join(constants)}")` since the f-string is  dynamically (double-quote) derived from a static `' '` (single-quoted) string. Plus in that latter example, it looks much nicer without the `\"` escaped jazz.

But totally understand your perspective too. Especially in .ipynb files, which aren't designed to be directly readable like .py files are. And .ipynb files are already cluttered with a bunch of syntactical jive such as `\n`s and `metadata{}`.

If there's time, I'll take a look at other pre-commits, such as flake8 or ruff, that can do most of what Black does, but with finer-grained control...

Thanks for the feedback. We can close this PR or leave it open--whichever you prefer. I'll keep working on this idea on my other fork/branch...

**Comment by giswqs** [comment]
I agree; the escape characters in the plain text of the .ipynb file can be overwhelming. However, the rendered notebook displays correctly without these characters. Alternatively, you can save the source code in a .ipy script. After applying the Black formatter, it will also appear clean and free of escape signs.

```python
print(f"Constants: {' '.join(constants)}")
print("Exceptions:", " ".join(exceptions["other"]))
print("Warnings:", " ".join(exceptions["warnings"]))
print("Errors:", " ".join(exceptions["errors"]))
print(f"Special: {' '.join(special_names)}")
print(f"Functions: {' '.join(functions)}")
print(f"Classes: {' '.join(classes)}")
print(f"Keywords: {' '.join(keywords)}")
```

**Comment by dotysan** [comment]
> ```python
> print(f"Constants: {' '.join(constants)}")
> print("Exceptions:", " ".join(exceptions["other"]))
> ```

Hold on. Now I'm confused. In one case `' '.join()` and on the next line `" ".join()`. Did you do that or Black?

**Comment by giswqs** [comment]
Yes, there are all formatted by black. I didn't do any manual changes

**Pull request closed without merging**